subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/PublishedDependencyConstraintsIntegrationTest.groovy

```diff
@@ -402,4 +404,61 @@ class PublishedDependencyConstraintsIntegrationTest extends AbstractModuleDepend
         }
     }
 
+    @RequiredFeatures(
+        [@RequiredFeature(feature = GradleMetadataResolveRunner.GRADLE_METADATA, value="true")]
+    )
+    void "deferred selector still resolved when constraint disappears"() {
+        repository {
+            'org:bar:1.0'()
+            'org:bar:1.1'()
+
+            'org:other:1.0' {
+                dependsOn 'org:bar:1.0'
+                dependsOn 'org:weird:1.1'
+            }
+
+            // Version 1.0 has a constraint, 1.1 does not
+            'org:weird:1.0' {
+                constraint 'org:bar:1.1'
+            }
+            'org:weird:1.1'()
+        }
+
+        buildFile << """
+dependencies {
+    conf 'org:weird:1.0'
+    conf 'org:other:1.0'
+}
+"""
+
+        repositoryInteractions {
```

Comment:
:o: I think it would be nice to formulate the exact interactions here. Also to document what is currently requested and what not. The other tests where you needed to adjust the `repositoryInteractions ` demonstrated that this is useful to see how changes like this PR influence the behavior of interacting with repositories.

Reply:
Good point, doing that!